### PR TITLE
`HTTPClient.perform`: changed response to `Result<[String: Any], Error>`

### PR DESCRIPTION
**Pull request by NachoSoto**
Currently pretty much all closures in the codebase use `(Value?, Error?)`, which is a leftover from Objective-C days. These can represent 4 different cases:
- **Value / no-error**
- **No-value / error**
- _Value / error_
- _No-value / no-error_

Out of those 4, only 2 are really valid, which meant that a lot of code was spent handling the other 2, either creating errors (like `UnexpectedBackendResponseSubErrorCode.loginMissingResponse`) that weren't possible, or hard to follow code because both a value and an error were possible.

In tests, too, the mocks were initially set up to produce no value and no error, which were in many cases triggering code paths that weren't legal with real code.

This change removes this possibility all together by ensuring **at compile time**, that one of the two is always present.

This is step 1 in a series of refactors.
- Step 2 will move _a lot of duplicated code_ that handles status codes != 20x, error handling, parsing error responses, etc. up to `HTTPClient` and out of the operations, making the operations a lot simpler.
- Step 3 will change `HTTPClient` to return a `Response: Decodable` instead of `[String: Any]`, moving more of the response serialization logic to `HTTPClient`, and leaving the operations to only focus on the "happy path".

### CODE REVIEW

**Comment by NachoSoto** [comment]
I'm seeing these failures locally too, but only sometimes 🤔 not when I run each test individually. It's failing to find the active `UIWindowScene`. I debugged this, and the scene is in "background" mode. 

I have an idea for how to improve this for tests by I don't understand what in this PR would make these fail all of a sudden?

**Comment by NachoSoto** [comment]
Tests fixed in #1390

File: Purchases/Networking/Operations/Handling/CustomerInfoResponseHandler.swift [hunk 1]
```diff
@@ -22,71 +22,73 @@ class CustomerInfoResponseHandler {
...
-        } else {
-            customerInfoError = nil
-            customerInfo = nil
-        }
+            var result: Result<CustomerInfo, Error> = {
+                // Only attempt to parse a response if we don't have an error status code from the backend.
+                if !isErrorStatusCode {
+                    return Result { try CustomerInfo.from(json: response) }
+                } else {
+                    let finishable = !statusCode.isServerError
```
Comment (NachoSoto) [inline Purchases/Networking/Operations/Handling/CustomerInfoResponseHandler.swift:46]: There's so much logic here that's going to go away and move to `HTTPClient`.

File: Purchases/Networking/Operations/Handling/SubscriberAttributeHandler.swift [hunk 2]
```diff
@@ -22,28 +22,28 @@ class SubscriberAttributeHandler {
...
-                                                    backendMessage: response["message"] as? String,
-                                                    extraUserInfo: extraUserInfo as [NSError.UserInfoKey: Any])
-        } else {
-            responseError = nil
-        }
-
-        completion(responseError)
+        let result: Result<[String: Any], Error> = response
+            .mapError {
+                ErrorUtils.networkError(withUnderlyingError: $0)
```
Comment (NachoSoto) [inline Purchases/Networking/Operations/Handling/SubscriberAttributeHandler.swift:29]: This line is basically repeated in all operations, so it'll be moved to `HTTPClient` too.

File: PurchasesTests/Purchasing/EntitlementInfosTests.swift [hunk 3]
```diff
@@ -472,23 +472,23 @@ class EntitlementInfosTests: XCTestCase {
...
+        subscriberInfo = try CustomerInfo(data: response)
         entitlement = try XCTUnwrap(subscriberInfo.entitlements.active["pro_cat"])
 
         expect(entitlement.ownershipType) == .familyShared
 
         stubResponse(entitlements: mockEntitlements,
                      subscriptions: mockSubscriptions(ownershipType: "BOATY_MCBOATFACE"))
 
-        subscriberInfo = try XCTUnwrap(CustomerInfo(testData: response))
+        subscriberInfo = try CustomerInfo(data: response)
```
Comment (NachoSoto) [inline PurchasesTests/Purchasing/EntitlementInfosTests.swift:491]: These are no longer using the test-only `.init(testData:)`. The actual method in the code `.init(data:)` already has a `throws` initializer, so don't need to use `XCTUnwrap`. I think using this we can probably get rid of `init(testData:)`.

File: PurchasesTests/Purchasing/PurchasesTests.swift [hunk 4]
```diff
@@ -454,33 +457,33 @@ class PurchasesTests: XCTestCase {
...
+        let customerInfoBeforePurchase = try CustomerInfo(data: [
             "request_date": "2019-08-16T10:30:42Z",
             "subscriber": [
                 "first_seen": "2019-07-17T00:05:54Z",
                 "original_app_user_id": "app_user_id",
                 "subscriptions": [:],
                 "non_subscriptions": [:]
             ]])
-        let customerInfoAfterPurchase = CustomerInfo(testData: [
+        let customerInfoAfterPurchase = try CustomerInfo(data: [
```
Comment (NachoSoto) [inline PurchasesTests/Purchasing/PurchasesTests.swift:477]: Same here, why have an `Optional` around when we can make sure at initialization that the data is correct.

File: StoreKitUnitTests/BeginRefundRequestHelperTests.swift [hunk 5]
```diff
@@ -270,49 +270,13 @@ class BeginRefundRequestHelperTests: XCTestCase {
...
-
-        do {
-            _ = try await helper.beginRefundRequestForActiveEntitlement()
-            XCTFail("beginRefundRequestForActiveEntitlement should have thrown error")
-        } catch {
-            expect(error).to(matchError(expectedError))
-            expect(error.localizedDescription) == expectedError.localizedDescription
-        }
-    }
-
```
Comment (NachoSoto) [inline StoreKitUnitTests/BeginRefundRequestHelperTests.swift:310]: Again testing something that wasn't possible.

File: PurchasesTests/Networking/HTTPClientTests.swift [hunk 6]
```diff
@@ -189,10 +189,9 @@ class HTTPClientTests: XCTestCase {
             response.error = error
             return response
         }
-        self.client.perform(request, authHeaders: [:]) { (status, data, responseError) in
-            if let responseNSError = responseError as NSError? {
+        self.client.perform(request, authHeaders: [:]) { (status, result) in
+            if let responseNSError = result.error as NSError? {
                 successFailed.value = (status.isServerError
```
Comment (aboedo) [inline PurchasesTests/Networking/HTTPClientTests.swift:194]: I know this variable isn't from this PR, but this name confuses me a lot - what does `successFailed` represent? 
Is it something like `operationFailed` or `requestFailed`?

File: PurchasesTests/Networking/HTTPClientTests.swift [hunk 7]
```diff
@@ -219,11 +218,10 @@ class HTTPClientTests: XCTestCase {
             )
         }
 
-        self.client.perform(request, authHeaders: [:]) { (status, data, error) in
-            correctResponse.value = (status.rawValue == errorCode) && (data != nil) && (error == nil)
-            if data != nil {
-                message.value = data?["message"] as? String
-            }
+        self.client.perform(request, authHeaders: [:]) { (status, result) in
+            correctResponse.value = (status.rawValue == errorCode) && (result.value != nil)
```
Comment (aboedo) [inline PurchasesTests/Networking/HTTPClientTests.swift:222]: [not for this PR] I feel like this variable would be more appropriately named `isResponseCorrect` or something that indicates it's a boolean, I kinda expected it to be an HTTPResponse instead

File: PurchasesTests/Networking/HTTPClientTests.swift [hunk 8]
```diff
@@ -292,11 +288,9 @@ class HTTPClientTests: XCTestCase {
                                      headers: nil)
         }
 
-        self.client.perform(request, authHeaders: [:]) { (status, data, error) in
-            successIsTrue.value = (status == .success) && (error == nil)
-            if data != nil {
-                message.value = data?["message"] as? String
-            }
+        self.client.perform(request, authHeaders: [:]) { (status, result) in
+            successIsTrue.value = (status == .success) && (result.error == nil)
```
Comment (aboedo) [inline PurchasesTests/Networking/HTTPClientTests.swift:292]: `isSuccessful`?

File: PurchasesTests/Purchasing/OfferingsManagerTests.swift [hunk 9]
```diff
@@ -168,31 +168,9 @@ extension OfferingsManagerTests {
         expect((error as NSError).code) == ErrorCode.unexpectedBackendResponseError.rawValue
     }
 
-    func testOfferingsForAppUserIDReturnsNilUnexpectedBackendResponseIfBackendReturnsNilDataAndNilOfferings() throws {
```
[the remark this answers is not in the document]
Comment (aboedo) [reply PurchasesTests/Purchasing/OfferingsManagerTests.swift:171]: sayonara

**Review by aboedo — COMMENTED** [review]
awesome!!! I love this refactor. I'd love to see pretty much everything in our codebase that requires completion blocks use Result, and this gets us like 90% of the way there 👏

File: Purchases/Purchasing/Purchases.swift [hunk 10]
```diff
@@ -748,12 +748,12 @@ public extension Purchases {
      */
     @objc(logIn:completion:)
     func logIn(_ appUserID: String, completion: @escaping (CustomerInfo?, Bool, Error?) -> Void) {
-        identityManager.logIn(appUserID: appUserID) { customerInfo, created, error in
+        identityManager.logIn(appUserID: appUserID) { result in
             self.operationDispatcher.dispatchOnMainThread {
-                completion(customerInfo, created, error)
+                completion(result.value?.info, result.value?.created ?? false, result.error)
```
[the remark this answers is not in the document]
Comment (NachoSoto) [reply Purchases/Purchasing/Purchases.swift:753]: Yeah exactly, but we could easily offer them too

File: Purchases/Identity/CustomerInfoManager.swift [hunk 11]
```diff
@@ -58,23 +60,24 @@ class CustomerInfoManager {
...
                                           isAppBackgrounded: Bool,
-                                          completion: ((CustomerInfo?, Error?) -> Void)?) {
+                                          completion: ((Result<CustomerInfo, Error>) -> Void)?) {
         let cachedCustomerInfo = cachedCustomerInfo(appUserID: appUserID)
         let isCacheStale = deviceCache.isCustomerInfoCacheStale(appUserID: appUserID,
                                                                 isAppBackgrounded: isAppBackgrounded)
-        let needsToRefresh = isCacheStale || cachedCustomerInfo == nil
-        if needsToRefresh {
+
+        guard !isCacheStale, let customerInfo = cachedCustomerInfo else {
```
[the remark this answers is not in the document]
Comment (NachoSoto) [reply Purchases/Identity/CustomerInfoManager.swift:68]: The reason I changed it is so that I could do `let customerInfo =` instead of `cachedCustomerInfo != nil` and then force-unwrapping :/

File: Purchases/Identity/IdentityManager.swift [hunk 12]
```diff
@@ -64,31 +64,29 @@ class IdentityManager {
...
             }
             return
         }
 
-        backend.logIn(currentAppUserID: currentAppUserID,
-                      newAppUserID: newAppUserID) { customerInfo, created, error in
-            if error == nil,
-               let customerInfo = customerInfo {
+        backend.logIn(currentAppUserID: currentAppUserID, newAppUserID: newAppUserID) { result in
+            if let customerInfo = result.value?.info {
```
[the remark this answers is not in the document]
Comment (NachoSoto) [reply Purchases/Identity/IdentityManager.swift:84]: Oh yeah much better

File: Purchases/Networking/Operations/GetOfferingsOperation.swift [hunk 13]
```diff
@@ -46,36 +46,26 @@ private extension GetOfferingsOperation {
...
+                .mapError { ErrorUtils.networkError(withUnderlyingError: $0) }
+                .flatMap { response in
+                    if statusCode.isSuccessfulResponse {
+                        return .success(response)
+                    } else {
+                        return .failure(
+                            ErrorUtils.backendError(withBackendCode: BackendErrorCode(code: response["code"]),
+                                                    backendMessage: response["message"] as? String)
+                        )
+                    }
```
[the remark this answers is not in the document]
Comment (NachoSoto) [reply Purchases/Networking/Operations/GetOfferingsOperation.swift:64]: Yup thank you!

File: Purchases/Networking/Operations/Handling/CustomerInfoResponseHandler.swift [hunk 14]
```diff
@@ -22,71 +22,73 @@ class CustomerInfoResponseHandler {
...
-            completion(nil, ErrorUtils.unexpectedBackendResponse(withSubError: customerInfoError,
-                                                                 extraContext: extraContext,
-                                                                 fileName: file, functionName: function, line: line))
-            return
-        }
+                    return .failure(
+                        ErrorUtils.backendError(withBackendCode: backendErrorCode,
+                                                backendMessage: message,
+                                                extraUserInfo: [
+                                                    ErrorDetails.finishableKey: finishable
```
[the remark this answers is not in the document]
Comment (NachoSoto) [reply Purchases/Networking/Operations/Handling/CustomerInfoResponseHandler.swift:54]: Yeah, at least in the next PR it'll be a bit more isolated to only a couple places.

File: Purchases/Networking/Operations/LogInOperation.swift [hunk 15]
```diff
@@ -51,58 +51,54 @@ private extension LogInOperation {
...
-        let result: (info: CustomerInfo?, cancelled: Bool, error: Error?) = {
-            if let error = error {
-                return (nil, false, ErrorUtils.networkError(withUnderlyingError: error))
+        let result: Result<(info: CustomerInfo, created: Bool), Error> = result
+            .mapError { ErrorUtils.networkError(withUnderlyingError: $0) }
+            .flatMap { response in
+                if !statusCode.isSuccessfulResponse {
+                    let backendCode = BackendErrorCode(code: response["code"])
+                    let backendMessage = response["message"] as? String
+                    let responsError = ErrorUtils.backendError(withBackendCode: backendCode,
```
[the remark this answers is not in the document]
Comment (NachoSoto) [reply Purchases/Networking/Operations/LogInOperation.swift:74]: What? There IS an _e_ 🙃 
Fixed.

File: Purchases/Purchasing/PurchasesOrchestrator.swift [hunk 16]
```diff
@@ -650,17 +622,16 @@ private extension PurchasesOrchestrator {
...
         if !self.allowSharingAppStoreAccount {
             Logger.warn(Strings.restore.restorepurchases_called_with_allow_sharing_appstore_account_false_warning)
         }
 
         let currentAppUserID = appUserID
         let unsyncedAttributes = unsyncedAttributes
         // Refresh the receipt and post to backend, this will allow the transactions to be transferred.
-        // swiftlint:disable line_length
+        // swiftlint:disable:next line_length
         // https://developer.apple.com/library/content/documentation/NetworkingInternet/Conceptual/StoreKitGuide/Chapters/Restoring.html
```
[the remark this answers is not in the document]
Comment (NachoSoto) [reply Purchases/Purchasing/PurchasesOrchestrator.swift:634]: Changed!

File: PurchasesTests/Networking/HTTPClientTests.swift (hunk 6 above)
Comment (NachoSoto) [reply PurchasesTests/Networking/HTTPClientTests.swift:194]: Changed! And I also changed the `expectToEventually` because it would wait _forever_ if the test fails instead of failing right away after this is evaluated.

File: PurchasesTests/Networking/HTTPClientTests.swift (hunk 7 above)
Comment (NachoSoto) [reply PurchasesTests/Networking/HTTPClientTests.swift:222]: Changed, and simplified the eventually calls too.

File: Purchases/Identity/CustomerInfoManager.swift (hunk 11 above)
[the remark this answers is not in the document]
Comment (aboedo) [reply Purchases/Identity/CustomerInfoManager.swift:68]: 👍  that makes sense
